### PR TITLE
Improved behavior on corrupted properties files

### PR DESCRIPTION
**Pull request by nijel**
When the properties file contains only key and lacks value, the
find_delimiter returns None as delimiter and it can be later serialized
as None string into the file. By setting delimiter to empty string, we
keep the file as is.

### CODE REVIEW

File: translate/storage/test_properties.py [hunk 1]
```diff
@@ -407,3 +407,12 @@ def test_joomla(self):
         assert propunit.name == u'VALUE'
         assert propunit.source == u'I am a "value"'
         assert bytes(propfile) == propsource
+
+    def test_serialize_missing_value(self):
+        propsource = 'key\n'.encode('utf-8')
```
Comment (unho) [inline translate/storage/test_properties.py]: When I read the PR description I was expecting the "no value" to mean that it was like `key=` but not like `key`. Please add a test for that case too.

**Comment by unho** [comment]
@nijel Besides the missing scenario raised in the added test, this is also lacking roundtrip testing with `prop2po` and `po2prop`.

File: translate/storage/test_properties.py (hunk 1 above)
Comment (nijel) [inline translate/storage/test_properties.py]: Maybe I should have described it as no delimiter rather :-).

**Comment by nijel** [comment]
I've added the test with missing value as well, but I don't see how this should be tested with prop2po - it does not store the delimiter, so there is no chance this would survive the roundtrip unchanged...

**Comment by unho** [comment]
> I've added the test with missing value as well, but I don't see how this should be tested with prop2po - it does not store the delimiter, so there is no chance this would survive the roundtrip unchanged...

That is exactly what we want to test, a test to check that PO has no-key and no-value and that final properties does have those. I guess that Weblate consumes properties directly and that is the reason why you need this. With po2prop this no-value and no-key should be added back if you provide a template properties with those no-key or no-value.

@nijel please note that **we need a working roundtrip**, and if it doesn't we have no other choice but hold this particular change until this can be done without breaking roundtrip. It is not that we want to impose you a certain way of doing things, but I want to stress the fact that there is people relying on roundtrip to work. I hope you will understand.

**Comment by nijel** [comment]
The roundtrip is certainly not working differently with the patch than before as all this changes is attribute which are not used in the conversion at all. I really don't see what you want to test there.

Anyway I've added two tests which sort of cover this, but still I don't see what you're expecting from them.

File: translate/convert/test_prop2po.py [hunk 2]
```diff
@@ -60,6 +60,15 @@ def test_convertprop(self):
         assert pounit.source == "Save file"
         assert pounit.target == ""
 
+    def test_novalue_entry(self):
```
Comment (unho) [inline translate/convert/test_prop2po.py]: This is `nokey`. Please add a test for `novalue` too.

Comment (unho) [inline translate/convert/test_prop2po.py]: I wanted to say this is `nodelimiter`.

File: translate/convert/test_po2prop.py [hunk 3]
```diff
@@ -62,6 +62,15 @@ def test_space_preservation(self):
         print(propfile)
         assert propfile == propexpected
 
+    def test_no_value(self):
```
Comment (unho) [inline translate/convert/test_po2prop.py]: This is `nodelimiter`. Please add a test for `novalue` too.

File: translate/convert/test_po2prop.py [hunk 4]
```diff
@@ -62,6 +62,15 @@ def test_space_preservation(self):
...
+        """check that we can handle keys without value"""
+        posource = '''#: KEY\nmsgctxt "KEY"\nmsgid ""\nmsgstr ""\n'''
+        proptemplate = '''KEY\n'''
+        propexpected = '''KEY = \n'''
```
Comment (unho) [inline translate/convert/test_po2prop.py]: `propexpected` is definitely different of `proptemplate`. This is the kind of inconsistency I wanted to add tests for, and it exposes that a change `po2prop` is needed to handle this.

Comment (unho) [inline translate/convert/test_po2prop.py]: Please check that output is the same as the template, and make the test xfail. I will be landing the needed fix in `po2prop` as it seems unfair to load you with the burden of fixing roundtrip.

**Comment by unho** [comment]
> The roundtrip is certainly not working differently with the patch than before as all this changes is attribute which are not used in the conversion at all. I really don't see what you want to test there.
>
>Anyway I've added two tests which sort of cover this, but still I don't see what you're expecting from them.

Increasing coverage is always a good thing, as this will prevent future messups. Roundtrip is working the same as always, but by introducing this new scenario it exposed a problem with roundtrip not handling correctly this scenario.

Please add the requested changes (add more tests, xfail them, adjust test names) and I will provide a lgtm once done (after another review), and once this PR is landed I will carry on with the require `po2prop` changes.

Thanks for your hard work on this, and sorry for being so picky.

**Comment by nijel** [comment]
Anything more needs to be fixed?

File: translate/convert/test_po2prop.py [hunk 5]
```diff
@@ -62,6 +64,25 @@ def test_space_preservation(self):
...
+        posource = '''#: KEY\nmsgctxt "KEY"\nmsgid ""\nmsgstr ""\n'''
+        proptemplate = '''KEY = \n'''
+        propexpected = '''KEY = \n'''
+        propfile = self.merge2prop(proptemplate, posource)
+        print(propfile)
+        assert propfile == propexpected
+
+    @mark.xfail(reason="This doesn't work as expected right now")
+    def test_no_separator(self):
+        """check that we can handle keys without value"""
```
Comment (unho) [inline translate/convert/test_po2prop.py]: Typo: *value*

**Comment by unho** [comment]
@nijel lgtm

**Pull request merged**